src/main/kotlin/com/exactpro/th2/codec/fixng/FixNgCodec.kt

```diff
@@ -221,19 +237,27 @@ class FixNgCodec(dictionary: IDictionaryStructure, settings: FixNgCodecSettings)
                     dictField
                 } else {
                     context.warning(DIRTY_MODE_WARNING_PREFIX + "Field does not exist in dictionary. Field tag: $tag. Field value: $value.")
-                    Primitive(false, tag.toString(), String::class.java, emptySet(), tag)
+                    Primitive(false, tag.toString(), emptyList(), String::class.java, emptySet(), tag)
                 }
             } else {
                 // we reached next part of the message
                 return@forEachField false
             }
 
-            field.decode(source, map, value, tag, isDirty, context)
+            field.decode(source, map, tagsSet, value, tag, isDirty, context)
             return@forEachField true
         }
 
         for (field in fields.values) {
-            if (field.isRequired && !map.contains(field.name) && field.name !in prereadHeaderFields) {
+            if (!field.isRequired) continue
+
+            val tag = when (field) {
+                is Primitive -> field.tag
+                is Group -> field.counter
+                else -> error("Only `Primitive` and `Group` fields expected to be `required`")
+            }
```

Comment:
suggestion: we should check dictionary on start and throw errors when structure is wrong.

Reply:
this is "error that never happens". We have `else` branch here because it required by Kotlin syntax.